### ISSUE TITLE
[FR] Method to OPEN Copilot in Note Area

### ISSUE THREAD

**Issue by Joolyan** [logancyang/obsidian-copilot#749]
The sidebar is usually too narrow for Copilot, so moving or opening Copilot in the note area provides the best experience.

Using the Commander plugin with the command `Copilot: Toggle Copilot Chat Window in Note Area` I've replaced the default ribbon icon with this action, but if Copilot is already open in another tab this command closes Copilot.

Yes, I could click to repeat the command which would re-open Copilot, but this becomes more complicated when trying to open Copilot from outside of Obsidian.

**Describe the solution you'd like**

I can use the URI `obsidian://search?vault=joolyan&query=` with [Quicksilver](https://qsapp.com/) to use a global keyboard shortcut to open/activate Obsidian with the cursor in the Search bar, so I can search Obsidian without using a mouse.

**Describe alternatives you've considered**

Using the `Advanced URI` plugin with the URI:  `obsidian://adv-uri?vault=<vault-name>&commandid=copilot%3Achat-toggle-window-note-area`
I can "toggle" Copilot, but not "open" Copilot. Also, using the URI doesn't move the cursor to the Copilot prompt window.

In summary, can you recommend an alternative method to set a global keyboard shortcut to open Copilot, or consider the feature request?

**Comment by logancyang** [comment]
I've also had this issue. At the very least, `Copilot: Toggle Copilot Chat Window in Note Area` should not close Copilot if it's already open on the side pane. So this should be fixed.

**Comment by logancyang** [comment]
Would you like to try this one? @zeroliu

**Comment by Joolyan** [comment]
Thanks for replying.

"Would you like to try this one?" - Sorry, which one? I'm currently using v2.7.0
Also, my use case is when using Copilot in the note view and not in the sidebar, which is usually hidden or too narrow. `Copilot: Toggle Copilot Chat Window in Note Area` currently closes Copilot when it's open in note view, whether it's currently the active tab or not.

Happy to test any solutions which may resolve this issue.

**Comment by zeroliu** [comment]
@logancyang I can give it a try.

@Joolyan to make sure I understand the issue, what's your expected behavior when copilot is already open in another tab when you trigger `Copilot: Toggle Copilot Chat Window in Note Area`. Would you expect it to focus on the copilot tab? If so, maybe it's better to build a different command (e.g. Copilot: Open Copilot Chat Window in Note Area). IMO, it's working as intended for "toggle" action to hide the tab if it's already on. This new command should also work with you advanced URI.

**Comment by logancyang** [comment]
> "IMO, it's working as intended" - I agree. The "Toggle" command works correctly which is why I raised the issue as a feature request. The command `Copilot: Open Copilot Chat Window in Note Area` would certainly help, but other solutions may include the option in settings to either use the sidebar or note view, and the new command `Open Copilot`. And there could be other solutions if the Copilot plugin included it's own URI scheme.
> 
> Does that help?

This is a great idea.

Initially I designed it to be a side pane because I'm on a bigger screen most of the time when I use Obsidian personally. When the FR for "toggle at the note area" I implemented the current logic which is kinda weird especially on mobile.

A setting for a default location, and consolidating the current 2 "toggle copilot" commands may be a more reasonable solution.

**Comment by Joolyan** [comment]
I also work mostly on large monitor, but may be because I switched to Obsidian from [Notational Velocity](https://notational.net/) then [nvALT](https://brettterpstra.com/projects/nvalt/), I prefer to keep my notes app UI as clean as possible. I even use the Auto Hide plugin to keep both sidebars closed, when not in use.

Using the note area may also make other features possible, such as the ability to maintain more than one chat conversation by having multiple Copilots open in different tabs. May be something for Copilot Plus?

Returning to this particular issue, may be the simplest solution is to add the commands `Copilot: Open Copilot Chat Window` and `Copilot: Open Copilot Chat Window in Note Area`. Also, ensuring the prompt dialog has focus when the Copilot window is opened. For some reason the focus is set correctly when the Toggle command is used within Obsidian, but not when used with the Advanced URI plugin.

Again, thanks for considering this feature request.

**Comment by zeroliu** [comment]
@Joolyan The FR is implemented in the attached PR. I added a new setting to configure whether to open the plugin in editor or in sidebar view. And you can use "Open Copilot Chat Window" command to open the plugin in the area based on the settings. The following advanced URI should also work.

`obsidian://advanced-uri?commandid=copilot%3Achat-open-window`

**Comment by zeroliu** [comment]
<img width="537" alt="Screenshot 2024-11-09 at 2 57 43 PM" src="https://github.com/user-attachments/assets/a32044ac-3d31-40df-bea1-54d6f9e9ceb7">

**Comment by Joolyan** [comment]
Many thanks for implementing this feature.

I've just tested with Copilot version 2.7.3 and the setting "Open Plugin In: Editor"...

Copilot adds the Ribbon icon "Copilot Chat" which behaves differently to the command "Copilot: Open Copilot Chat Window"

Copilot Chat (Ribbon icon)
Pro's: Opens Copilot with curser in the chat window
Con's: Closes Copilot if already open

Command "Copilot: Open Copilot Chat Window" using Advanced URI
Pro's: Doesn't close the Copilot window if already open
Con's: Requires Shift+Tab, Shift+Tab, Shift+Tab to move curser to the chat window

Using Commander, it's possible to hide the default Copilot Chat ribbon icon and add an icon for the command "Copilot: Open Copilot Chat Window".

Any idea why, when using Advanced URI, the cursor isn't in the chat window?

**Comment by zeroliu** [comment]
https://github.com/logancyang/obsidian-copilot/pull/824

This PR should fix the focus issue.

The sidebar icon is bound to toggle window instead of open window. It's easy to change to open command but I will let @logancyang to make a final call.

**Comment by zeroliu** [comment]
I just tested my change with hotkeys and realized that the open window command will not mess up with the layout because we disable and re-enable the plugin. I will follow up with a new fix for this use case. @Joolyan in case it affects your use case too

**Comment by Joolyan** [comment]
@zeroliu Thanks for testing.

"we disable and re-enable the plugin" - this explains some of the screen re-draw behaviour I'm seeing when using the "Open Copilot Chat Window" command. 

FYI My use case is to use Obsidian Copilot using only the keyboard. This is now possible, but hoping the focus fix will result in the cursor appearing in the chat window (ready to type the text prompt) when the "Open Copilot Chat Window" command is called via a URI.

**Comment by zeroliu** [comment]
This PR should fix it: https://github.com/logancyang/obsidian-copilot/pull/845

**Comment by zeroliu** [comment]
`obsidian://advanced-uri?commandid=copilot%3Achat-open-window` should behave exactly like what you described. In 2.7.4, it will cause a redraw. The next release should only focus on your opened chat view!

**Comment by Joolyan** [comment]
Thanks and look forward to testing.
